Fix conditinal request expiration

meggiman opened the pull request:
Closes #464 

- Update `CacheMixin`s `_send_and_cache` function to save CachedResponse with updated expiration date upon received 304 before returning the cached value.
- Update `BaseCache`s `save_response` function to only create a `CachedResponse` instance if the argument is not already of type `CachedResponse`

JWCook commented on requests_cache/backends/base.py:
```diff
@@ -92,8 +92,12 @@ def save_response(self, response: AnyResponse, cache_key: str = None, expires: d
         cache_key = cache_key or self.create_key(response.request)
-        cached_response = CachedResponse.from_response(response, expires=expires)
-        cached_response = redact_response(cached_response, self.ignored_parameters)
+        if isinstance(response, CachedResponse):
```
Could we keep this section the way it is, and put this check in `CachedResponse.from_response` instead? E.g.:
```python
    @classmethod
    def from_response(cls, original_response: Response, expires: datetime = None, **kwargs):
        """Create a CachedResponse based on an original response object"""
        if isinstance(original_response, CachedResponse):
            original_response.expires = expires
            return original_response

        obj = cls(expires=expires, **kwargs)

        ...
```

meggiman replied:
Done. Although I opted to return a copy of the original_response instead of the same object. I guess the user would not expect to be served the same instance.

JWCook commented:
Could you add a unit test for these changes? Also, looks like the test for immediate expiration (`max-age=0` plus a validator) is currently failing.

meggiman commented:
Thanks a lot for the review! I implemented your suggestions with a few twists:
- Instead of just returning the original CachedResponse object from the "from_response()" class method I create a copy of the object using attr.evolve. I guess users would not expect builder class methods like from_response to return the very same object but create a new one that can be modified without affecting the original.
- According to the RFC mentioned in #464, we should update the `actions` variable using the combination of cached response and "Cache-Control" headers present in the 304 response. I thus added another `actions.update_from_response()` call if the response is in fact a `304 Not Modified`.
- I added a unit test that verifies the behavior explained above (i.e. the server relaxing `max-age` with its 304 response from 0 to 60s) and makes sure the Session doesn't try to re-validate for another `max-age` seconds. 

With the changes outlined above, the regression on the `test_conditional_request__max_age_0` test should have been resolved as well.

JWCook commented:
Merged. Thanks for the contribution!

Looks like tests for python 3.9 are failing, but I believe that's an issue with the build environment and not related to your changes. I'll fix that.

JWCook commented:
@meggiman Your changes are now in pre-release version `0.9.0dev2`!

meggiman commented:
Awesome! Thanks a lot for being so quick to react to this issue and for this awesome project in general. It really saved me from a lot of trouble!
